### PR TITLE
network: add support to dualstack

### PR DESCRIPTION
**Pull request by dougsland**
The dualstack introduces apiVIPs and ingressVIPs and deprecate
apiVIP and ingressVIP fields. This change just adapt the script.

Signed-off-by: Douglas Schilling Landgraf <dougsland@redhat.com>

### CODE REVIEW

**Comment by dougsland** [comment]
Tested using @creydr patch https://github.com/openshift/installer/pull/5798 and seems to work.

File: ocp_install_env.sh
Comment (creydr) [inline ocp_install_env.sh:144]: When https://github.com/openshift/installer/pull/5798 got merged, you can also specify the single vip in the plural field. (-> no need for `apiVIP` and `ingressVIP` anymore). E.g.:
```
    apiVIPs:
    - apiVIP
    ingressVIPs:
    - ingressVIP
```
This allows you to have the same rendering for one and multiple vips, probably too.

File: ocp_install_env.sh
Comment (cybertron) [reply ocp_install_env.sh:139]: Yeah, this won't do what we need. The comma format isn't valid for YAML.

File: ocp_install_env.sh
Comment (cybertron) [reply ocp_install_env.sh:144]: We still need support for older releases in dev-scripts so we need to have a way to use the old field names.

However, I'd prefer to look at the openshift release number to determine which we need. Even in the case of a single VIP in 4.12, we should still be using the new fields (the current implementation would always use the old fields for single VIP deployments, IIUC). At some point the deprecated ones will go away and we'd have to revisit this anyway. Better to do it now when it's fresh in our minds.

File: ocp_install_env.sh
Comment (cybertron) [inline ocp_install_env.sh:133]: Nit: this isn't specific to dual stack. This function will be called for all deployment types.

File: ocp_install_env.sh
Comment (cybertron) [inline ocp_install_env.sh:136]: This makes a couple of assumptions I'm not sure are valid:

- That if there are multiple API VIPs there will always be multiple ingress VIPs
- That we only want to use the new fields if multiple VIPs are specified

I think this needs to be based on the openshift version like it is for dnsvip, and the logic should handle a mismatch in the number of VIPs between ingress and API. I'm not sure the latter is strictly necessary, but there's no inherent reason you couldn't have two API VIPs and only one ingress (or vice versa).

**Review by cybertron — CHANGES_REQUESTED** [review]

**Comment by dougsland** [comment]
/unhold looks like it's working

Here some quick demo out of devscripts for testing:
```
#!/bin/bash

function set_multiple_resources() {
    # This function helps to write in a YAML format multiple resources.
    # Example: apiVIPs and ingressVIPs
    resource="${2}";
    echo "    ${1}"
    for data in ${resource//,/ }; do
        echo "    - ${data}"
    done
}

function openshift_version() {
    return 4
}

function set_VIPs() {
    # apiVIP and ingressVIP both has been DEPRECATED in 4.12 in favor of apiVIPs and ingressVIPs.
    # This functions helps to write the new apiVIPs/ingressVIPs format or set the old fields.
    case "${1}" in

    apivips)
        #if printf '%s\n4.12\n' "$(openshift_version)" | sort -V -C; then
            if [[ "${API_VIP}" == *","* ]]; then
                set_multiple_resources "apiVIPs:" "${API_VIP}"
                return
            fi
        #else
# fallback mode for users not using the new apiVIPs schema
cat <<EOF
    apiVIP: ${API_VIP}
EOF
        #fi
    ;;
    ingressvips)
        #if printf '%s\n4.12\n' "$(openshift_version)" | sort -V -C; then
            if [[ "${INGRESS_VIP}" == *","* ]]; then
                set_multiple_resources "ingressVIPs:" "${INGRESS_VIP}"
                return
            fi
        #else
# fallback mode for users not using the new igressVIPs schema
cat <<EOF
     ingressVIP: ${INGRESS_VIP}
EOF
        #fi
    ;;
    esac
}


API_VIP="192.168.2.3"
INGRESS_VIP="192.168.1.1, 192.168.2.3"
set_VIPs apivips
set_VIPs ingressvips

```

File: ocp_install_env.sh
Comment (creydr) [inline ocp_install_env.sh:167]: same as above

**Review by creydr — CHANGES_REQUESTED** [review]

File: ocp_install_env.sh
Comment (dougsland) [reply ocp_install_env.sh:149]: sure

File: ocp_install_env.sh
Comment (creydr) [reply ocp_install_env.sh:151]: But isn't the fallback not for the version check?

File: ocp_install_env.sh
Comment (creydr) [inline ocp_install_env.sh:149]: needs to be intended (same level in yaml as `apiVIPs` - as you did in line 159)

File: ocp_install_env.sh
Comment (creydr) [inline ocp_install_env.sh:153]: same as above

**Review by creydr — CHANGES_REQUESTED** [review]
In case of OCP version <= 4.11 and having multiple VIPs defined in API_VIP this would result in picking only the first one. IMHO we should render the full string (instead of only the first VIP) and then later let the validation fail on it as the user entered garbage (multiple VIPs not supported in <=4.11).
And I am wondering if this could be written without playing around with `IFS`, but this might be a personal thing :D

**Comment by dougsland** [comment]
> In case of OCP version <= 4.11 and having multiple VIPs defined in API_VIP this would result in picking only the first one. IMHO we should render the full string (instead of only the first VIP) and then later let the validation fail on it as the user entered garbage (multiple VIPs not supported in <=4.11). And I am wondering if this could be written without playing around with `IFS`, but this might be a personal thing :D

All good,  adjusted the script.

File: ocp_install_env.sh
Comment (creydr) [inline ocp_install_env.sh:157]: nit: as you need `echo` in renderVIPs too, you could use `echo "    apiVIP: ${API_VIP}"` here to... Probably makes it clearer to read here.

File: ocp_install_env.sh
Comment (creydr) [inline ocp_install_env.sh:168]: same as above

**Comment by creydr** [comment]
Seems lgtm from my side. @cybertron could you check on this too?

**Comment by vpickard** [comment]
@cybertron Can you please give ok-to-test on this PR? Thanks

**Review by cybertron — COMMENTED** [review]
Looks like a good start. In my local tests the new logic seems to do the right thing.

We do need a followup, either in this PR or a separate one, to actually populate both VIPs when dual stack is in use. I think that's going to require the following things:

- Adding a second DNS entry for the api in vm_setup_vars.yaml (this may need to be conditionalized so we don't do it in single-stack deployments)
- Adding a dual stack conditional in network.sh that looks up the new api DNS entry and adds it to $API_VIP.
- Ditto for ingress, except it's a simple variable assignment there
- Add secondary entries to the NM dnsmasq config in network.sh

I'm not sure how much the unmanaged br bridge case gets used these days, so we can probably leave that side of the logic alone. If someone needs dual stack vip support there we can work with them to do that, but in the meantime I'm not sure we have a good way to test it anyway.

**Comment by zhaozhanqi** [comment]
@cybertron can we merge this PR?

**Comment by cybertron** [comment]
The current version of this patch is only a partial implementation. We still need to actually configure both VIPs when dev-scripts is run in dual stack mode. Right now you have to hack in overrides for the VIP variables.

**Comment by mkowalski** [comment]
Drive-by review, but is there a chance of e.g. providing a sample of `config.sh` with a correct syntax to use dual-stack VIPs? It would really help other teams that use dev-scripts and don't have internal knowledge of the implementation

**Comment by cybertron** [comment]
> Drive-by review, but is there a chance of e.g. providing a sample of `config.sh` with a correct syntax to use dual-stack VIPs? It would really help other teams that use dev-scripts and don't have internal knowledge of the implementation

The intent is that when this is complete you'll get dual VIPs any time you set IP_STACK=v4v6. The VIP variables aren't actually exposed as an interface so I don't think it's possible to set them manually.

**Comment by dougsland** [comment]
tested locally, `ipv4 and dualstack` everything `working`.
@zhaozhanqi  @cybertron @creydr  ready for review and tests.

Thanks!

**Comment by dougsland** [comment]
Please see the commit message for details about this feature enabler and testing.

**Comment by dougsland** [comment]
Tested in the following scenarios and everything worked:
```
OPENSHIFT 4.11
    - v4   - [x]
    - v4v6 - [x]
    - v6   - [x]

OPENSHIFT 4.12
    - v4   - [x]
    - v4v6 - [x]
    - v6   - [x]
```

**Comment by dougsland** [comment]
I don't believe this is related to the patch:
```
ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.\n\nWe recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.\n\nrequests-oauthlib 1.3.1 requires oauthlib>=3.0.0, but you'll have oauthlib 2.1.0 which is incompatible.\nflask-oauthlib 0.9.6 requires requests-oauthlib<1.2.0,>=0.6.2, but you'll have requests-oauthlib 1.3.1 which is incompatible.\n",
```

**Review by cybertron — CHANGES_REQUESTED** [review]
Mostly just a few nitpicks, but I think we should fix the redundant version logic because that's going to be confusing. One of those if branches is dead code.

File: network.sh
Comment (dougsland) [reply network.sh:219]: /me nods. Probably good idea to rename the function.

File: network.sh
Comment (dougsland) [reply network.sh:253]: sure

File: ocp_install_env.sh
Comment (dougsland) [reply ocp_install_env.sh:127]: done

**Comment by cybertron** [comment]
Waiting on ci, but this lgtm now.

**Comment by dougsland** [comment]
> Waiting on ci, but this lgtm now.

I am also waiting all my local tests finish here, so far, so good.

**Comment by dougsland** [comment]
CI error not related to patch:
```
    "requirements": null,
    "state": "present",
    "stderr": "WARNING: Running pip install with root privileges is generally not a good idea. Try `pip3 install --user` instead.\nERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.\n\nWe recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.\n\nrequests-oauthlib 1.3.1 requires oauthlib>=3.0.0, but you'll have oauthlib 2.1.0 which is incompatible.\nflask-oauthlib 0.9.6 requires requests-oauthlib<1.2.0,>=0.6.2, but you'll have requests-oauthlib 1.3.1 which is incompatible.\n",
    "stderr_lines": [
        "WARNING: Running pip install with root privileges is generally not a good idea. Try `pip3 install --user` instead.",
        "ERROR: After October 2020 you may experience errors when installing or updating packages. This is because pip will change the way that it resolves dependency conflicts.",
        "",
        "We recommend you use --use-feature=2020-resolver to test your packages with the new resolver before it becomes the default.",
        "",
        "requests-oauthlib 1.3.1 requires oauthlib>=3.0.0, but you'll have oauthlib 2.1.0 which is incompatible.",
        "flask-oauthlib 0.9.6 requires requests-oauthlib<1.2.0,>=0.6.2, but you'll have requests-oauthlib 1.3.1 which is incompatible."
```

**Comment by dougsland** [comment]
all local tests worked with the following config:

```
OPENSHIFT 4.11
    - v4   - [x]
    - v4v6 - [x]
    - v6   - [x]

OPENSHIFT 4.12
    - v4   - [x]
    - v4v6 - [x]
    - v6   - [x]
```

**Review by creydr — COMMENTED** [review]
/lgtm

**Comment by cybertron** [comment]
> /lgtm
> 
> Tested with `IP_STACK=v4v6`: 4.12: worked 4.11: Got error message, as expected with invalid install config (2 vips in apiVIP / ingressVIP field)

/hold

We can't break 4.11 dual stack with this patch. It looks like we need to make the version check smart enough to only include the first VIP on older releases.

**Comment by dougsland** [comment]
already merged in another PR.

**Pull request closed without merging**
